### PR TITLE
Controls: Add button control

### PR DESCRIPTION
**Pull request by Markcial**
Issue: https://github.com/storybookjs/storybook/issues/11971

Adds button control to addon-control

## What I did

I added a new button control on the controls add-on

## How to test

- Is this testable with Jest or Chromatic screenshots? Don't know, will follow reviewers advice
- Does this need a new example in the kitchen sink apps? Yes, Will need to update some examples
- Does this need an update to the documentation? Yes, will add after getting the draft accepted

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

File: lib/components/src/controls/Button.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  },
+}));
+
+export type ButtonProps = ControlProps<ButtonValue> & ButtonConfig;
+export const ButtonControl: FC<ButtonProps> = ({
+  name,
+  label = name, // FIXME: Use name if label is missing?
+  defaultValue,
+  value = defaultValue,
+  onChange, // FIXME: how to use this?
```
Comment (yngvebn) [inline lib/components/src/controls/Button.tsx:44]: Didn't think buttons had onChange events? Can't this just be removed?

Comment (Markcial) [reply lib/components/src/controls/Button.tsx:44]:
```suggestion
```

**Review by shilman — COMMENTED** [review]
Hey @Markcial! Thanks for picking this up!

The code is good/clean, but there is a subtlety that's worth discussing before merging, namely Storybook's channel-based architecture and what it means for button actions per this PR.

## Problem

Specifically, controls can show up in two different places: (1) the controls addon panel, (2) the docs page. The former is located in Storybook's "manager" UI, and the latter is located in Storybook's "preview" iframe. The manager and the preview communicate across a channel, and everything that goes on the channel should be [telejson-serializable](https://github.com/storybookjs/telejson). All user code (stories + components) is executed in the preview iframe.

What this means is that I would expect:
- Simple functions like your demo would work in both controls & docs
- Complex functions that reference any preview state would only work in docs

This could be very confusing for users, depending on what they are trying to accomplish. "Why does button action A work but B fails?"

## Solution

One possible solution is to always execute the button action in the preview, so it is guaranteed to have all the context it needs and not rely on brittle serialization logic.

1. The control would store a global mapping from button ID to action
2. Clicking the button would dispatch a "button action" event on the channel
3. A handler would receive the event in the preview, look up the action function in the mapping from (1), and execute it.

What do you think?

**Comment by Markcial** [comment]
I totally agree, I knew nothing about this.

Is there any test I can grab as a boilerplate? Seeing that this is easy to miss I would prefer to include testing to avoid regressions, and it also would help me to speed up the changes

Any test that does that, for example sending a message after changing a select on the controls or similar. I can just adapt it to this specific use case.

Besides that, the design needs review or are we ok with just that copy and paste?

Thanks a lot for taking some time on this

**Comment by shilman** [comment]
Re: design I'm hoping @ghengeveld can look it over when he's back next week

Re: reference implementation, I'm not sure of the best example to look at. I think the `addon-kit` round-trip is a pretty good example of using the channel: https://github.com/storybookjs/addon-kit/blob/main/src/Panel.js#L23

**Comment by Markcial** [comment]
I added the implementation requested, it's sketchy but works.

Tell me if some changes are needed, because I might not be using the right standards or placing the files right

And If you point me to the proper places I can start adding documentation and examples

**Comment by yannbf** [comment]
Hey @Markcial that looks interesting. Could you show an example that would work the same as with this knobs example?

```jsx
export const ButtonWithReactUseState = () => {
  const [counter, setCounter] = React.useState(0);
  button('increment', () => setCounter(counter + 1));
  button('decrement', () => setCounter(counter - 1));
  return <>Value is changed once you click in the knobs button: {counter}</>;
};
```

**Comment by Markcial** [comment]
@yannbf  That's a tricky question. Because the way controls work I can't handle state, it should be provided to the component. Anyways I can work out something similar if the component it's managed through the props. I will commit a new story today with an example.

**Comment by Markcial** [comment]
I tried, and I got something close to work, but unfortunately the args get updated after some more emit calls. I am posting this commit for informational sake on what was my idea, but seems to be that maybe I need to think a bit and use a different approach.

Any insight or clue on what I can do to circumvent this limitation it's welcome, I will keep thinking on how to add this functionality on the meanwhile, thanks

**Comment by yannbf** [comment]
> I tried, and I got something close to work, but unfortunately the args get updated after some more emit calls. I am posting this commit for informational sake on what was my idea, but seems to be that maybe I need to think a bit and use a different approach.
> 
> Any insight or clue on what I can do to circumvent this limitation it's welcome, I will keep thinking on how to add this functionality on the meanwhile, thanks

Yea, I think the declarative nature of controls makes it quite difficult to achieve something that before was simple with knobs. That makes me wonder if it's something we want to support, and if so, maybe create a good standard for it. WDYT @shilman ?

**Comment by ghengeveld** [comment]
@shilman and I discussed this PR just now. I brought up the concern that Args are supposed to be used as props that go into the component. A button that just invokes some random function in the context of the story, but doesn't directly affect the args, should not be passed as an arg itself. Storybook has the concept of Parameters for the purpose of passing metadata into the story, and so that should be used instead.

Besides the technical implementation, we have concerns about the conceptual aspects as well. Currently, the Controls addon is designed to handle Args, not Parameters (or Globals, another similar concept). Adding this Button control would increase the scope of the Controls addon to include a whole new concept. That might prove necessary and worthwhile, but really we need to take a step back and reconsider the use cases we're trying to address here.

Based on this PR, I see two use cases being addressed:
  1. Invoking a function in the scope of the rendered story, potentially to trigger interactions on the component
  2. Dynamically mutating args at the click of a button, potentially as a way to have arg "presets" or "macros"

The first use case is part of a project we're going to work on over the summer. One feature we intend to add is a `setup` function that will be invoked immediately after rendering the story, so that it can be used to bring your component into a state that cannot be reached by setting args, parameters or props. Essentially this is meant to be used for stateful components. The first use case would be an extension of this, where a similar function is invoked based on an action in the Storybook UI. We'd prefer to have a dedicated UI and API for this functionality rather than hijacking Controls and Args/ArgTypes for this purpose.

The second use case is directly involved with mutating args, so this might realistically be part of Controls. However, we again would prefer to have a dedicated UI and API for this purpose. Either a new Macros addon, or a dedicated section in the Controls UI comes to mind. If this thing were to also handle Parameters and Globals, I think a separate Macros addon makes more sense, to keep Controls focussed on Args alone.

@Markcial To determine a path forward for this PR, and your involvement in the larger project, we have to clarify which of these use cases you are specifically worried about. Can you describe the ways in which you intended to use this Button control? This is really important for us to design something that works well for the use case rather than creating a one-size-fit-nobody solution.

I hope you don't feel bad about this. Rest assured we're committed to work it out, hopefully with your continued involvement. We have to keep in mind the bigger picture, which requires careful deliberation before accepting changes we might come to regret.

**Comment by Markcial** [comment]
Hi, thanks for this clarifications.

Thing is that I already solved that by using a different approach, so if you close the issue because you feel that is no relevant or there is no user base for this I won't feel bad at all.

I was trying to contribute on what back on the time sounded like a good idea. But you are the maintainers and I will totally respect your decision to reject a badly designed idea or something that has no return value at all.

But if you find that there is some path for me to help on this I am eager to help whenever I get some spare time.

As you are the ones that have to carry the bigger burden feel free to close it, modify it, or tell me what I should do to readdress this PR to something more usable

Thanks

**Comment by ndelangen** [comment]
@shilman Sounds like the way forward with this is to close it? (for now?)

**Comment by shilman** [comment]
I just deprecated addon-knobs and I expect to see some kind of reaction after we release 6.3 publicly. Let's keep this around until 6.3 stabilizes.

**Comment by edgardz** [comment]
In my case I have some components that expose methods with useImperativeHandle hook, like "animateIn()" and "animateOut()". Button controls would be very handy to test these methods.

**Comment by ndelangen** [comment]
6.3 feels stabilized to me @shilman

**Comment by MichaelArestad** [comment]
@shilman Do you think we should close this now?

**Pull request closed without merging**